pkg/securitypolicy/securitypolicyenforcer_rego.go

```diff
@@ -21,50 +36,491 @@ func init() {
+	options := ast.CompileOpts{
+		EnablePrintStatements: false,
+	}
+
+	if compiled, err := ast.CompileModulesWithOpt(modules, options); err == nil {
+		policy.compiledModules = compiled
+	} else {
+		return nil, fmt.Errorf("rego compilation failed: %w", err)
+	}
+
+	return policy, nil
+}
+
+func (policy RegoEnforcer) Query(input map[string]interface{}) (rego.ResultSet, error) {
+	store := inmem.NewFromObject(policy.data)
+
+	var buf bytes.Buffer
+	rule := input["name"].(string)
+	query := rego.New(
+		rego.Query(fmt.Sprintf("data.policy.%s", rule)),
+		rego.Compiler(policy.compiledModules),
+		rego.Input(input),
+		rego.Store(store),
+		rego.PrintHook(topdown.NewPrintHook(&buf)))
+
+	ctx := context.Background()
+	results, err := query.Eval(ctx)
+	if err != nil {
+		log.G(ctx).WithError(err).WithFields(logrus.Fields{
+			"Policy": policy.objects,
+		})
+		return results, err
+	}
+
+	output := buf.String()
+	if len(output) > 0 {
+		log.G(ctx).Debug(output)
+	}
+
+	return results, nil
+}
+
+func (policy *RegoEnforcer) EnforceDeviceMountPolicy(target string, deviceHash string) error {
+	policy.mutex.Lock()
+	defer policy.mutex.Unlock()
+
+	input := map[string]interface{}{
+		"name":       "mount_device",
+		"target":     target,
+		"deviceHash": deviceHash,
+	}
+	result, err := policy.Query(input)
+	if err != nil {
+		return err
+	}
+
+	if !result.Allowed() {
+		input_json, err := json.Marshal(input)
+		if err != nil {
+			return fmt.Errorf("Unable to marshal the Rego input data.")
+		}
+
+		return fmt.Errorf("device mount not allowed by policy.\ninput: %s", string(input_json))
+	}
+
+	deviceMap := policy.data["devices"].(map[string]string)
+	if _, found := deviceMap[target]; found {
+		input_json, err := json.Marshal(input)
+		if err != nil {
+			return fmt.Errorf("Unable to marshal the Rego input data.")
+		}
+
+		return fmt.Errorf("device %s already mounted.\ninput: %s", target, string(input_json))
+	}
+
+	deviceMap[target] = deviceHash
+	return nil
+}
+
+func (policy *RegoEnforcer) EnforceOverlayMountPolicy(containerID string, layerPaths []string) error {
+	policy.mutex.Lock()
+	defer policy.mutex.Unlock()
+
+	input := map[string]interface{}{
+		"name":        "mount_overlay",
+		"containerID": containerID,
+		"layerPaths":  layerPaths,
+	}
+	result, err := policy.Query(input)
+	if err != nil {
+		return err
+	}
+
+	if !result.Allowed() {
+		input_json, err := json.Marshal(input)
+		if err != nil {
+			return fmt.Errorf("Unable to marshal the Rego input data.")
+		}
+
+		return fmt.Errorf("overlay mount not allowed by policy.\ninput: %s", string(input_json))
+	}
+
+	// we store the mapping of container ID -> layerPaths for later
+	// use in EnforceCreateContainerPolicy here.
+	containerMap := policy.data["containers"].(map[string]interface{})
+	if _, found := containerMap[containerID]; found {
+		input_json, err := json.Marshal(input)
+		if err != nil {
+			return fmt.Errorf("Unable to marshal the Rego input data.")
```

Comment:
Idiomatic Go generally has errors start lowercase and not end with punctuation. Also `errors.New` would be preferred as no formatting needed

Reply:
I've updated all these.

Reply:
Thanks Sean!